src/widgets.js

```diff
@@ -292,7 +296,69 @@ define([
         return widget;
     };
 
-    
+    var dropdown = function (mug, options) {
+        var widget = normal(mug, options);
+        widget.input = $("<select />")
+            .attr("name", widget.id)
+            .addClass('input-block-level');
+
+        var input = widget.input;
+
+        widget.setValue = function (value, option_name) {
+            var val = equivalentOption(value);
+            if (val) {
+                input.val(val.value);
+            } else {
+                widget.addOption(value, option_name || "");
+                input.val(value);
+            }
+        };
+
+        widget.getValue = function () {
+            return input.val();
+        };
+
+        widget.updateValue = function () {
+            widget.save();
+        };
+
+        input.change(function () {
+            widget.handleChange();
+        });
+
+        widget.addOption = function (value, text) {
+            var option = $('<option />')
+                .attr('value', value)
+                .text(text);
+            this.input.append(option);
+        };
+
+        widget.addOptions = function (options) {
+            var _this = this;
+            _.forEach(options, function(option) {
+                _this.addOption(option.value, option.text);
+            });
+        };
+
+        widget.getOptions = function () {
+            return _.map(widget.input.find('option'), function(option) {
+                return {
+                    value: option.value,
+                    text: option.text
+                };
+            });
+        };
+
+        function equivalentOption(val) {
+            val = val ? JSON.parse(val) : '';
+            return _.find(widget.getOptions(), function (option) {
+                return _.isEqual(option.value ? JSON.parse(option.value) : '', val);
```

Comment:
widget.getOptions returns a dictionary of {value: value, text: text} so this is right.